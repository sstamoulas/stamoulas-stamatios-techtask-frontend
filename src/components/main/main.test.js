import React from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import ReactTestUtils, { act } from 'react-dom/test-utils'
import { waitForElement, screen } from '@testing-library/react'

import Main from './main.component'
import RecipesActionTypes from './../../redux/recipes/recipes.types'

const mockStore = configureStore([])
const mealDate = new Date('07/14/2020')
const ingredients = [{
  "title": "Cheese",
  "use-by": "2020-01-08"
},
{
  "title": "Bread",
  "use-by": "2020-11-01"
},
{
  "title": "Butter",
  "use-by": "2020-11-25"
}]
const isFetching = false
const selectedIngredients = []

describe('Main Component', () => {
  let store
  let container
 
  beforeEach(() => {
    store = mockStore({
      ingredients: {
        ingredients,
      },
      mealDate: {
        mealDate,
      },
      recipes: {
        isFetching,
        selectedIngredients
      },
    })

    store.dispatch = jest.fn()

    container = document.createElement('div')
    document.body.appendChild(container)

    act(() => {
      ReactDOM.render(
        <Provider store={store}>
          <Main />
        </Provider>, 
        container
      )
    })
  })

  it('should render with given state from Redux store with no selectedIngredients', () => {
    expect(container).toMatchSnapshot()
  })

  it('should render with given state from Redux store with atleast one selectedIngredient', () => {
    store.getState().recipes.selectedIngredients = ['Bread']
    expect(container).toMatchSnapshot()
  })

  it('should dispatch an action when an ingredient is selected', async () => {
    ReactTestUtils.Simulate.click(container.querySelector('input.rw-input-reset'))
    
    await waitForElement(() => screen.getByText(/Bread/i))

    ReactTestUtils.Simulate.click(screen.getByText(/Bread/i))

    expect(container.querySelector('.rw-multiselect-tag span').textContent).toContain('Bread')

    selectedIngredients.push('Bread')

    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(store.dispatch).toHaveBeenCalledWith({ 
      type: RecipesActionTypes.FETCH_RECIPES_START, 
      payload: selectedIngredients 
    })
  })
})
