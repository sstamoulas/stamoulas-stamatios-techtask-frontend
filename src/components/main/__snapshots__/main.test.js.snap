// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Component should render with given state from Redux store with atleast one selectedIngredient 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      <div
        class="input-container"
      >
        <label
          for="date"
        >
          Meal Date:  
        </label>
        <input
          id="date"
          type="date"
          value="2020-07-14"
        />
      </div>
      <div
        class="select-container"
      >
        <div
          class="rw-multiselect rw-widget"
          tabindex="-1"
        >
          <span
            aria-atomic="true"
            aria-live="assertive"
            aria-relevant="additions removals text"
            class="rw-sr"
            id="rw_2_notify_area"
            role="status"
          />
          <div
            class="rw-widget-input rw-widget-picker rw-widget-container"
          >
            <div>
              <input
                aria-busy="false"
                aria-disabled="false"
                aria-expanded="false"
                aria-haspopup="true"
                aria-owns="rw_2_listbox rw_2_notify_area "
                aria-readonly="false"
                autocomplete="off"
                class="rw-input-reset"
                placeholder=""
                role="listbox"
                size="2"
                tabindex="0"
                value=""
              />
            </div>
            <span
              class="rw-select"
            >
              <button
                aria-disabled="false"
                aria-hidden="true"
                class="rw-btn rw-btn-select"
                role="presentational"
                tabindex="-1"
                type="button"
              />
            </span>
          </div>
        </div>
      </div>
      <div
        class="card-container"
      >
        <div>
          There are no recipes to display
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Main Component should render with given state from Redux store with no selectedIngredients 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      <div
        class="input-container"
      >
        <label
          for="date"
        >
          Meal Date:  
        </label>
        <input
          id="date"
          type="date"
          value="2020-07-14"
        />
      </div>
      <div
        class="select-container"
      >
        <div
          class="rw-multiselect rw-widget"
          tabindex="-1"
        >
          <span
            aria-atomic="true"
            aria-live="assertive"
            aria-relevant="additions removals text"
            class="rw-sr"
            id="rw_1_notify_area"
            role="status"
          />
          <div
            class="rw-widget-input rw-widget-picker rw-widget-container"
          >
            <div>
              <input
                aria-busy="false"
                aria-disabled="false"
                aria-expanded="false"
                aria-haspopup="true"
                aria-owns="rw_1_listbox rw_1_notify_area "
                aria-readonly="false"
                autocomplete="off"
                class="rw-input-reset"
                placeholder=""
                role="listbox"
                size="2"
                tabindex="0"
                value=""
              />
            </div>
            <span
              class="rw-select"
            >
              <button
                aria-disabled="false"
                aria-hidden="true"
                class="rw-btn rw-btn-select"
                role="presentational"
                tabindex="-1"
                type="button"
              />
            </span>
          </div>
        </div>
      </div>
      <div
        class="card-container"
      >
        <div>
          There are no recipes to display
        </div>
      </div>
    </div>
  </div>
</div>
`;
